File: audio2face-2d/nodejs/scripts/audio2face-2d.js
// Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
//
// Permission is hereby granted, free of charge, to any person obtaining a
// copy of this software and associated documentation files (the "Software"),
// to deal in the Software without restriction, including without limitation
// the rights to use, copy, modify, merge, publish, distribute, sublicense,
// and/or sell copies of the Software, and to permit persons to whom the
// Software is furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
// DEALINGS IN THE SOFTWARE.

// Include/import modules
const assert = require('assert');
const fs = require('fs');
const grpc = require('@grpc/grpc-js');
const http = require('http');
const messages = require('../interfaces/audio2face2d_pb')
const memoryStream = require('memorystream');
const path=require("path");
const services = require('../interfaces/audio2face2d_grpc_pb');
const writer = require('wav').Writer;
const { Command } = require('commander');
const { parse } = require('csv-parse/sync')

// Set parameters
const chunkSize = 1024 * 1024 // Size of audio chunk to stream
const memStream = new memoryStream(); // To stream video output from server to the browser
const A2F2D_BROWSER_PORT = process.env.A2F2D_BROWSER_PORT || 3000; // check if system has env variable for browser streaming PORT, otherwise use 3000
 
/**
 * Sets up the video streaming server for streaming output video to browser
 *
 * @returns {handle} handle for the server object
 */
function setupVideoStreamingServer() {
  // Create the server
  const server = http.createServer((req, res)=>{
    if(req.method === 'GET' && req.url === "/"){
      // Send index.html page when user visits "/" endpoint
      // index.html will have video component that displays the video
      fs.createReadStream(path.resolve("index.html")).pipe(res);
      return;
    }

    //if video content is requested
    if(req.method === 'GET' && req.url === "/video"){
      const head = {
	      'Content-Type': 'video/mp4',
      }
      res.writeHead(200, head);
      memStream.pipe(res);
    }
    else{
      res.writeHead(400);
      res.end("bad request");
    }
  })

  return server;
}

/**
 * Processes the head pose mode animation input data from the provided file paths.
 * 
 * @param {string} headRotationAnimationFilepath - The path to the head rotation animation input file.
 * @param {string} headTranslationAnimationFilepath - The path to the head translation animation input file.
 * 
 * @returns {object} An object containing the processed rotation and translation data streams.
 */
function processHeadPoseModeAnimationInput(headRotationAnimationFilepath, headTranslationAnimationFilepath) {
  // Check if head rotation and translation input files exist
  if (!fs.existsSync(headRotationAnimationFilepath)) {
    console.error('Head rotation animation input file %s does not exist', headRotationAnimationFilepath);
    process.exit(1); // Exit the process with a non-zero status code
  }
  if (!fs.existsSync(headTranslationAnimationFilepath)) {
    console.error('Head translation animation input file %s does not exist', headTranslationAnimationFilepath);
    process.exit(1); // Exit the process with a non-zero status code
  }
  
  // Setup the head rotation and translation animation
  const rotation = fs.readFileSync(headRotationAnimationFilepath, 'utf8');
  const translation = fs.readFileSync(headTranslationAnimationFilepath, 'utf8');

  // Parse the CSV data
  const rotationCsv = parse(rotation, {
    delimiter: ',',
    columns: false,
    cast: (value) => parseFloat(value),
  });
  const translationCsv = parse(translation, {
    delimiter: ',',
    columns: false,
    cast: (value) => parseFloat(value),
  });

  // Convert data to float32
  const rotationData = rotationCsv.map(row => row.map(Number));
  const translationData = translationCsv.map(row => row.map(Number));

  // Assertions
  assert(Array.isArray(rotationData) && Array.isArray(rotationData[0]), 'Data should be a 2-dimensional array');
  assert(rotationData.length > 0, 'Data should have more than 0 rows');
  rotationData.forEach(row => {
    assert(row.length === 4, 'Each row should have 4 columns');
  });
  assert(Array.isArray(translationData) && Array.isArray(translationData[0]), 'Data should be a 2-dimensional array');
  assert(translationData.length > 0, 'Data should have more than 0 rows');
  translationData.forEach(row => {
    assert(row.length === 3, 'Each row should have 3 columns');
  });

  // Create rotation and translation data streams
  const rotationDataStream = new messages.QuaternionStream({ values: [] });
  const translationDataStream = new messages.Vector3fStream({ values: [] });

  // Populate the rotation and translation data streams
  for (const x of rotationData) {
    var q = new messages.Quaternion();
    q.setX(x[0]);
    q.setY(x[1]);
    q.setZ(x[2]);
    q.setW(x[3]);
    rotationDataStream.addValues(q);
  }

  for (const x of translationData) {
    var v = new messages.Vector3f();
    v.setX(x[0]);
    v.setY(x[1]);
    v.setZ(x[2]);
    translationDataStream.addValues(v);
  }

  return {rotationDataStream, translationDataStream};
}

/**
 * Sends the input config like portrait image, mode, model etc
 *
 * @param rpcCall - RPC call object
 * @param {string} portraitInputFile - Path to the portrait input file
 * @param {string} headRotationAnimationFilepath - Path to the rotation animation data file
 * @param {string} headTranslationAnimationFilepath - Path to the translation animation data file
 */
function sendInputConfig(rpcCall, portraitInputFile, headRotationAnimationFilepath, headTranslationAnimationFilepath) {
  // Open and read the portrait image in binary format
  const config = new messages.AnimateConfig();
  const imgData = fs.readFileSync(portraitInputFile);
  try {
    config.setPortraitImage(imgData);
  }
  catch (err) {
    console.error('Error reading the image file:', err);
    process.exit(1); // Exit the process with a non-zero status code
  }

  // Prepare the config for Audio2Face 2D
  const headPoseMode = messages.HeadPoseMode.HEAD_POSE_MODE_RETAIN_FROM_PORTRAIT_IMAGE;
  config.setAnimationCropMode(messages.AnimationCroppingMode.ANIMATION_CROPPING_MODE_REGISTRATION_BLENDING);
  config.setModelSelection(messages.ModelSelection.MODEL_SELECTION_QUALITY);
  config.setEnableLookaway(1);
  config.setLookawayMaxOffset(20);
  config.setLookawayIntervalMin(240);
  config.setLookawayIntervalRange(90);
  config.setBlinkFrequency(15);
  config.setBlinkDuration(6);
  config.setMouthExpressionMultiplier(1.4);
  config.setHeadPoseMode(headPoseMode);
  config.setHeadPoseMultiplier(0.4);

  if (headPoseMode == messages.HeadPoseMode.HEAD_POSE_MODE_USER_DEFINED_ANIMATION) {
    // Process the head pose rotation and translation animation data from input CSV files
    var animationData = processHeadPoseModeAnimationInput(headRotationAnimationFilepath, headTranslationAnimationFilepath);

    // Set the Audio2face 2D config with head rotation and animation data
    config.setInputHeadRotation(animationData.rotationDataStream);
    config.setInputHeadTranslation(animationData.translationDataStream);
  }

  var msg = new messages.AnimateRequest();
  msg.setConfig(config);

  // Write the config to the stream
  try {
    rpcCall.write(msg);
    console.log("Input config sent");
  } catch (writeError) {
    console.error('Write Error:', writeError.message);
    process.exit(1); // Exit the process with a non-zero status code
  }
}

/**
 * Sends input audio chunks
 *
 * @param rpcCall - RPC call object
 * @param format - type of audio format (pcm or wav)
 * @param {string} audioInputFile - Path to audio input file
 */
function sendInputAudioChunks(rpcCall, format, audioInputFile) {
  // Read audio file
  let readStream;
  if (format === 'wav') {
    readStream = fs.createReadStream(audioInputFile, { highWaterMark: chunkSize });
  } else if (format === 'pcm') {
    const pcm = fs.createReadStream(audioInputFile);

    // Convert pcm format to wav 
    // Change the pcm config here, if needed
    readStream = new writer({
      sampleRate: 48000,
      channels: 1,
      bitDepth: 16
    });

    pcm.pipe(readStream);
  }

  readStream.on('data', (chunk) => {
    // Process the chunk here
    var msg = new messages.AnimateRequest();
    msg.setAudioFileData(chunk);

    // Write the audio chunk to the stream
    rpcCall.write(msg);
  });

  readStream.on('end', () => {
    rpcCall.end();
    console.log('Input audio sent');
    console.log('Receiving and writing the video chunks from server to the output file');
  });

  readStream.on('error', (err) => {
    console.error('Error reading the audio file:', err);
    process.exit(1); // Exit the process with a non-zero status code
  });
}

/**
 * Sets up the infra to receive response(video chunks) from the server
 *
 * @param rpcCall - RPC call object
 * @param {string} outputFile - Path to video output file
 * @param {bool} browser - Specifies if the browser is enabled
 */
function receiveServerResponse(rpcCall, outputFile, browser) {
  // Open the mp4 file to write the video frames from the server
  const writeStream = fs.createWriteStream(outputFile);
  
  // Set up processing for the response
  rpcCall.on('data', function(response) {
    if (response.hasVideoFileData()) {
      const data = response.getVideoFileData();
      writeStream.write(data);

      if (browser) {
        memStream.write(data);
      }
    }
  });

  rpcCall.on('end', () => {
    writeStream.end();

    if (browser) {
      memStream.end();
    }

    // Log the elapsed time
    console.timeEnd('Function invocation completed in ');
    console.log(`video written to ${outputFile}`);

    if (browser) {
      console.log("Press Ctrl+C to stop the browser streaming server");
    }
  });

  rpcCall.on('error', (err) => {
    console.error('Error:', err);
    process.exit(1); // Exit the process with a non-zero status code
  });
}

function readFileContent(filePath) {
  /**
   * Function to read file content as bytes.
   *
   * @param {string} filePath - Path to the input file
   * @returns {Buffer} - File content as bytes
   * @throws {Error} - If the file cannot be read
   */
  try {
    return fs.readFileSync(filePath); // Reads the file content as a Buffer
  } catch (err) {
    console.error(`Error reading file at ${filePath}:`, err.message);
    throw err;
  }
}

/**
 * Parses the command line arguments
 *
 * @returns {options} options available from command-line
 */
function parseArguments() {
  const program = new Command();

  program
    .description('Node JS client for A2F2D NIM')
    .option('--ssl-mode <type>', 'Flag to set SSL mode, default is DISABLED', 'DISABLED')
    .option('--ssl-key <type>', 'The path to SSL private key.', '../ssl_key/ssl_key_client.pem')
    .option('--ssl-cert <type>', 'The path to SSL certificate chain.', '../ssl_key/ssl_cert_client.pem')
    .option('--ssl-root-cert <type>', 'The path to SSL root certificate.', '../ssl_key/ssl_ca_cert.pem')
    .option('--target <type>', 'IP:port of gRPC service, when hosted locally.', '127.0.0.1:8001')
    .option('--audio-input <type>', 'The path to the input audio file.', '../../assets/sample_audio.wav')
    .option('--head-rotation-animation-filepath <type>', 'The path to head rotation animation csv file.', '../../assets/head_rotation_animation.csv')
    .option('--head-translation-animation-filepath <type>', 'The path to head translation animation csv file.', '../../assets/head_translation_animation.csv')
    .option('--portrait-input <type>', 'The path to the input portrait file.', '../../assets/sample_portrait_image.png')
    .option('--format <type>', 'Audio format - wav or pcm', 'wav')
    .option('--browser', 'Specifies whether to enable browser streaming')
    .option('--output <type>', 'The path for the output video file.', 'output.mp4')
    .helpOption('-h, --help', 'Display help for command');

  program.parse(process.argv);
  return program.opts();
}

/**
 * Creates a gRPC client instance based on the provided options.
 * 
 * @param {object} options - The options object containing the following properties:
 *   - sslMode {string}: The SSL mode to use for the connection. Can be one of 'DISABLED', 'TLS', or 'MTLS'.
 *   - sslRootCert {string}: The path to the root certificate file (required for 'TLS' and 'MTLS' modes).
 *   - sslKey {string}: The path to the private key file (required for 'MTLS' mode).
 *   - sslCert {string}: The path to the certificate chain file (required for 'MTLS' mode).
 *   - target {string}: The target URL for the gRPC client.
 * 
 * @returns {object} The created gRPC client instance.
 * 
 * @throws {Error} If the provided SSL mode is invalid or if required certificate/key files are missing.
 */
function createGrpcClient(options) {
  let client;
  let credentials;

  switch (options.sslMode) {
    case 'DISABLED':
      console.log('Using insecure connection');
      credentials = grpc.credentials.createInsecure();
      break;

    case 'TLS':
      if (!options.sslRootCert) {
        console.error('If --ssl-mode is TLS, --ssl-root-cert is required.');
        process.exit(1);
      }
      const rootCert = readFileContent(options.sslRootCert);
      credentials = grpc.credentials.createSsl(rootCert);
      console.log('Using secure connection with TLS');
      break;

    case 'MTLS':
      if (!(options.sslKey && options.sslCert && options.sslRootCert)) {
        console.error(
          'If --ssl-mode is MTLS, --ssl-key, --ssl-cert, and --ssl-root-cert are required.'
        );
        process.exit(1);
      }
      const privateKey = readFileContent(options.sslKey);
      const certificateChain = readFileContent(options.sslCert);
      const rootCertificates = readFileContent(options.sslRootCert);
      credentials = grpc.credentials.createSsl(rootCertificates, privateKey, certificateChain);
      console.log('Using secure connection with MTLS');
      break;

    default:
      console.error(`Invalid SSL mode: ${options.sslMode}`);
      process.exit(1);
  }

  // Create the gRPC client
  client = new services.Audio2Face2DServiceClient(options.target, credentials);
  return client;
}

function main() {
  // Parse the commandline arguments
  // Note: commander package used for parsing arguments internally changes all options with dashes to camel case
  // For example audio-input to audioInput. Therefore, all the options with dashes will be referenced with camelcase notation
  const options = parseArguments();

  // Validate SSL mode
  const validSslModes = ['DISABLED', 'MTLS', 'TLS'];
  if (!validSslModes.includes(options['sslMode'])) {
    console.error(`Invalid SSL mode. Must be one of: ${validSslModes.join(', ')}`);
    process.exit(1);
  }

  // Validate the audio format
  if (options['format'] != 'pcm' && options['format'] != 'wav') {
    console.error('Audio format must be pcm or wav');
    process.exit(1); // Exit the process with a non-zero status code
  }

  // Validate the provided audio input file for correct format
  var ext = options['audioInput'].slice(((options['audioInput'].lastIndexOf(".") - 1) >>> 0) + 2);
  if (options['format'] == 'wav' && ext != 'wav' ||
    options['format'] == 'pcm' && ext != 'raw') {
      console.error('Input audio file format must match the audio format: file format (extension): %s, audio format: %s', ext, options['format']);
      process.exit(1); // Exit the process with a non-zero status code
  }

  // Check if required input files exist
  if (!fs.existsSync(options['audioInput'])) {
    console.error('Audio input file %s does not exist', options['audioInput']);
    process.exit(1); // Exit the process with a non-zero status code
  }
  if (!fs.existsSync(options['portraitInput'])) {
    console.error('Portrait input file %s does not exist', options['portraitInput']);
    process.exit(1); // Exit the process with a non-zero status code
  }

  // Setup and start the video streaming server, if enabled
  if (options['browser']) {
    const server = setupVideoStreamingServer();
    server.listen(A2F2D_BROWSER_PORT , () => {
      console.log(`Server listening on port:${A2F2D_BROWSER_PORT }`);
    });
  }

  // Set up the connection with the server
  var client = createGrpcClient(options);
  
  // Start the timer
  console.time('Function invocation completed in ');

  // Setup call to the remote function (animate is the remote procedure for A2F2D)
  const rpcCall = client.animate((error) => {
    if (error) {
      console.error('RPC call error ', error.message);
      process.exit(1); // Exit the process with a non-zero status code
    }
  });

  // Check for connection error
  rpcCall.on('error', (error) => {
    console.error('Connection failed:', error.message);
    process.exit(1); // Exit the process with a non-zero status code
  });

  // Write the Audio2face 2D config
  sendInputConfig(rpcCall, options['portraitInput'], options['headRotationAnimationFilepath'], options['headTranslationAnimationFilepath']);

  // Write the audio chunks of size chunkSize from the audio file
  sendInputAudioChunks(rpcCall, options['format'], options['audioInput']);
  
  // Sets up receiving response from server and writing to a file
  receiveServerResponse(rpcCall, options['output'], options['browser']);
}

main()

File: audio2face-2d/python/scripts/audio2face-2d.py
# Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
#
# Permission is hereby granted, free of charge, to any person obtaining a
# copy of this software and associated documentation files (the "Software"),
# to deal in the Software without restriction, including without limitation
# the rights to use, copy, modify, merge, publish, distribute, sublicense,
# and/or sell copies of the Software, and to permit persons to whom the
# Software is furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
# DEALINGS IN THE SOFTWARE.

import argparse
import os
import sys
import time
import io
import grpc

sys.path.append(os.path.join(os.getcwd(), "../interfaces"))
# Importing gRPC compiler auto-generated maxine audio2face-2d library
import audio2face2d_pb2  # noqa: E402
import audio2face2d_pb2_grpc  # noqa: E402
from audio2face2d_pb2 import (  # noqa: E402
    QuaternionStream,
    Quaternion,
    Vector3fStream,
    Vector3f,
    ModelSelection,
    AnimationCroppingMode,
    HeadPoseMode,
)


def parse_args() -> None:
    """
    Parse command-line arguments using argparse.
    """
    # Set up argument parsing
    parser = argparse.ArgumentParser(
        description="Process input audio and portrait files and apply audio2face-2d effect."
    )
    parser.add_argument(
        "--ssl-mode",
        type=str,
        help="Flag to set SSL mode, default is None",
        default="DISABLED",
        choices=["DISABLED", "MTLS", "TLS"],
    )
    parser.add_argument(
        "--ssl-key",
        type=str,
        default="../ssl_key/ssl_key_client.pem",
        help="The path to ssl private key.",
    )
    parser.add_argument(
        "--ssl-cert",
        type=str,
        default="../ssl_key/ssl_cert_client.pem",
        help="The path to ssl certificate chain.",
    )
    parser.add_argument(
        "--ssl-root-cert",
        type=str,
        default="../ssl_key/ssl_ca_cert.pem",
        help="The path to ssl root certificate.",
    )
    parser.add_argument(
        "--target",
        type=str,
        default="127.0.0.1:8001",
        help="IP:port of gRPC service, when hosted locally.",
    )
    parser.add_argument(
        "--audio-input",
        type=str,
        default="../../assets/sample_audio.wav",
        help="The path to the input audio file.",
    )
    parser.add_argument(
        "--portrait-input",
        type=str,
        default="../../assets/sample_portrait_image.png",
        help="The path to the input portrait file.",
    )
    parser.add_argument(
        "--output",
        type=str,
        default="output.mp4",
        help="The path for the output video file.",
    )
    parser.add_argument(
        "--head-rotation-animation-filepath",
        type=str,
        default="../../assets/head_rotation_animation.csv",
        help="The path for the head_rotation_animation.csv file. "
        "Only required for HEAD_POSE_MODE_USER_DEFINED_ANIMATION",
    )
    parser.add_argument(
        "--head-translation-animation-filepath",
        type=str,
        default="../../assets/head_translation_animation.csv",
        help="The path for the head_translation_animation.csv file. "
        "Only required for HEAD_POSE_MODE_USER_DEFINED_ANIMATION",
    )
    return parser.parse_args()


def read_file_content(file_path: os.PathLike) -> None:
    """Function to read file content as bytes.

    Args:
      file_path: Path to input file
    """
    with open(file_path, "rb") as file:
        return file.read()


def generate_request_for_inference(audio_filepath: str, params: dict):
    """Generator to produce the request data stream

    Args:
      audio_filepath: Path to input file
      params: Parameters for the feature
    """
    yield audio2face2d_pb2.AnimateRequest(config=audio2face2d_pb2.AnimateConfig(**params))
    file = open(audio_filepath, "rb")
    while True:
        buffer = file.read(1024 * 1024)
        if buffer == b"":
            break
        yield audio2face2d_pb2.AnimateRequest(audio_file_data=buffer)
    print("Data sending done")


def process_head_pose_data(head_rotation_path, head_translation_path):
    """
    Process head rotation and translation data.

    Args:
        head_rotation_path (str): Path to the head rotation animation file.
        head_translation_path (str): Path to the head translation animation file.

    Returns:
        Tuple[QuaternionStream, Vector3fStream]: Processed rotation and translation data streams.
    """
    # Read the head rotation data
    with io.StringIO(open(head_rotation_path, "rb").read().decode("utf-8")) as file:
        head_rotation_data = []
        for line in file:
            values = line.strip().split(",")
            if len(values) == 4:
                head_rotation_data.append([float(val) for val in values])

    # Validate the data
    assert len(head_rotation_data) > 0, "Head rotation data is empty"
    assert all(len(row) == 4 for row in head_rotation_data), "Each row must have 4 values"

    # Create the QuaternionStream
    rotation_data_stream = QuaternionStream()
    for x in head_rotation_data:
        q = Quaternion()
        q.x, q.y, q.z, q.w = x
        rotation_data_stream.values.append(q)

    # Read the head translation data
    with io.StringIO(open(head_translation_path, "rb").read().decode("utf-8")) as file:
        head_translation_data = []
        for line in file:
            values = line.strip().split(",")
            if len(values) == 3:
                head_translation_data.append([float(val) for val in values])

    # Validate the data
    assert len(head_translation_data) > 0, "Head translation data is empty"
    assert all(len(row) == 3 for row in head_translation_data), "Each row must have 3 values"

    # Create the Vector3fStream
    translation_data_stream = Vector3fStream()
    for x in head_translation_data:
        v = Vector3f()
        v.x, v.y, v.z = x
        translation_data_stream.values.append(v)

    return rotation_data_stream, translation_data_stream


def process_request(
    channel: any,
    audio_filepath: os.PathLike,
    params: dict,
    output_filepath: os.PathLike,
) -> None:
    """Function to process gRPC request

    Args:
      channel: gRPC channel for server client communication
      input_filepath: Path to input file
      params: Parameters to control the feature
      output_filepath: Path to output file
      request_metadata: Credentials to process preview request
    """
    try:
        stub = audio2face2d_pb2_grpc.Audio2Face2DServiceStub(channel)
        start_time = time.time()
        responses = stub.Animate(
            generate_request_for_inference(audio_filepath=audio_filepath, params=params)
        )
        next(responses)
        file = open(output_filepath, "wb")
        print(f"Writing output in {output_filepath}")
        for response in responses:
            if response.HasField("video_file_data"):
                file.write(response.video_file_data)
        end_time = time.time()
        print(
            f"Function invocation completed in {end_time-start_time:.2f}s, "
            f"{output_filepath} file is generated."
        )
    except Exception as e:
        print(f"An error occurred: {e}")


def main():
    """
    Main client function
    """
    args = parse_args()
    portrait_filepath = args.portrait_input
    audio_filepath = args.audio_input
    output_filepath = args.output

    # Check file path
    if os.path.isfile(portrait_filepath):
        print(f"The image file '{portrait_filepath}' exists. Checking for audio file.")
    else:
        raise FileNotFoundError(f"The image file '{portrait_filepath}' does not exist. Exiting.")
    if os.path.isfile(audio_filepath):
        print(f"The audio file '{audio_filepath}' exists. Proceeding with processing.")
    else:
        raise FileNotFoundError(f"The audio file '{audio_filepath}' does not exist. Exiting.")

    portrait_image_encoded = open(portrait_filepath, "rb").read()

    # Configure head pose mode
    head_pose_mode = HeadPoseMode.HEAD_POSE_MODE_RETAIN_FROM_PORTRAIT_IMAGE

    # Provide head pose animation values for head pose mode HEAD_POSE_MODE_USER_DEFINED_ANIMATION
    if head_pose_mode == HeadPoseMode.HEAD_POSE_MODE_USER_DEFINED_ANIMATION:
        rotation_data_stream, translation_data_stream = process_head_pose_data(
            args.head_rotation_animation_filepath,
            args.head_translation_animation_filepath,
        )

    # Supply params as shown below, refer to the docs for more info.
    feature_params = {
        "portrait_image": portrait_image_encoded,
        "model_selection": ModelSelection.MODEL_SELECTION_QUALITY,
        "animation_crop_mode": AnimationCroppingMode.ANIMATION_CROPPING_MODE_REGISTRATION_BLENDING,
        "enable_lookaway": 1,  # can be 0 or 1
        "lookaway_max_offset": 20,  # value in [5, 25]
        "lookaway_interval_min": 240,  # value in [1, 600]
        "lookaway_interval_range": 90,  # value in [1, 600]
        "blink_frequency": 15,  # value in [0, 120]
        "blink_duration": 6,  # value in [2, 150]
        "mouth_expression_multiplier": 1.4,  # value in [1.0, 2.0]
        "head_pose_mode": head_pose_mode,
        "head_pose_multiplier": 1.0,  # value in [0.0, 1.0]
        # "input_head_rotation": rotation_data_stream,  # HEAD_POSE_MODE_USER_DEFINED_ANIMATION
        # "input_head_translation": translation_data_stream, # HEAD_POSE_MODE_USER_DEFINED_ANIMATION
    }

    # Check ssl-mode and create channel_credentials for that mode
    if args.ssl_mode != "DISABLED":
        channel_credentials = ""
        if args.ssl_mode == "MTLS":
            if not (args.ssl_key and args.ssl_cert and args.ssl_root_cert):
                raise RuntimeError(
                    "If --ssl-mode is MTLS, --ssl-key, --ssl-cert and --ssl-root-cert are required."
                )
            private_key = read_file_content(args.ssl_key)
            certificate_chain = read_file_content(args.ssl_cert)
            root_certificates = read_file_content(args.ssl_root_cert)
            channel_credentials = grpc.ssl_channel_credentials(
                root_certificates=root_certificates,
                private_key=private_key,
                certificate_chain=certificate_chain,
            )
        else:
            if not (args.ssl_root_cert):
                raise RuntimeError("If --ssl-mode is TLS, --ssl-root-cert is required.")
            root_certificates = read_file_content(args.ssl_root_cert)
            channel_credentials = grpc.ssl_channel_credentials(root_certificates=root_certificates)

        # Establish secure channel when ssl-mode is MTLS/TLS
        with grpc.secure_channel(target=args.target, credentials=channel_credentials) as channel:
            process_request(
                channel=channel,
                audio_filepath=audio_filepath,
                params=feature_params,
                output_filepath=output_filepath,
            )
    else:
        # Establish insecure channel when ssl-mode is DISABLED
        with grpc.insecure_channel(target=args.target) as channel:
            process_request(
                channel=channel,
                audio_filepath=audio_filepath,
                params=feature_params,
                output_filepath=output_filepath,
            )


if __name__ == "__main__":
    main()
